File: binary-search-tree-iterator/binary-search-tree-iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: TreeNode):
        self.root = root
        self.stack = []
        self.buildStack(self.root)

    def next(self) -> int:
        '''
        Collects the value in the current node and moves the pointer
        Returns the value in the current node
        '''
        rootval = self.stack.pop()
        if rootval.right:
            self.buildStack(rootval.right)
        return rootval.val

    def hasNext(self) -> bool:
        '''
        Checks if an element remains to right of the current pointer
        Returns true if it exists, false otherwise
        '''
        if self.stack:
            return True
        else:
            return False
        
    def buildStack(self, root: TreeNode):
        '''
        Builds the stack to traverse the BST
        '''
        while root:
            self.stack.append(root)
            if root.left:
                root = root.left
            else:
                root = None


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: binary-search-tree-iterator/binary-search-tree-iterator.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator 
{

    private List<Integer> inorder;
    private int location;
    
    public BSTIterator(TreeNode root) 
    {
        // Flatten the tree to perform iteration
        inorder = new ArrayList<>();
        DFS(root);
        location = -1;
    }
    
    private void DFS(TreeNode root)
    {
        // No subtree exists, return
        if (root == null) return;
        
        // Perform inorder DFS traversal
        DFS(root.left);
        inorder.add(root.val);
        DFS(root.right);
    }
    
    public int next() 
    {
        // If location in tree bounds, return values
        if (location+1 < inorder.size())
        {
            location += 1;
            return inorder.get(location);
        }
        return -1;
    }
    
    public boolean hasNext() 
    {
        // If location in bounds, return true
        if (location+1 < inorder.size())
            return true;
        return false;
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */